sdk/include/platform/sunburst/platform-adc.hh

```diff
@@ -0,0 +1,350 @@
+		ArduinoA3   = static_cast<uint8_t>(RegisterOffset::VoltageAuxiliary13),
+		ArduinoA4   = static_cast<uint8_t>(RegisterOffset::VoltageAuxiliary6),
+		ArduinoA5   = static_cast<uint8_t>(RegisterOffset::VoltageAuxiliary14),
+		Temperature = static_cast<uint8_t>(RegisterOffset::Temperature),
+		VoltageInternalSupply =
+		  static_cast<uint8_t>(RegisterOffset::VoltageInternalSupply),
+		VoltageAuxiliarySupply =
+		  static_cast<uint8_t>(RegisterOffset::VoltageAuxiliarySupply),
+		VoltageInternalReferencePositive = static_cast<uint8_t>(
+		  RegisterOffset::VoltageInternalPositiveReference),
+		VoltageInternalReferenceNegative = static_cast<uint8_t>(
+		  RegisterOffset::VoltageInternalNegativeReference),
+		VoltageBlockRamSupply =
+		  static_cast<uint8_t>(RegisterOffset::VoltageBlockRamSupply),
+	};
+
+	/**
+	 * Possible power down modes that can be set in Config Register 2.
+	 *
+	 * https://docs.amd.com/r/en-US/ug480_7Series_XADC/Control-Registers?section=XREF_93518_Power_Down
+	 */
+	enum class PowerDownMode : uint8_t
+	{
+		None = 0b00, /* Default */
+		/* 0b01 is not valid, and should not be selected. */
+		ConverterB     = 0b10,
+		BothConverters = 0b11,
+	};
+
+	/**
+	 * The Xilinx Analogue-to-Digtal Converter can sample at a maximum rate of 1
+	 * Megasample per second. It uses 16 bit Dynamic Reconfiguration Port (DRP)
+	 * registers, and stores 12-bit measurements, which are stored most-
+	 * significant-bit justified.
+	 *
+	 * https://docs.amd.com/r/en-US/ug480_7Series_XADC/XADC-Overview
+	 */
+	static constexpr size_t MaxSamples          = 1 * 1000 * 1000;
+	static constexpr size_t RegisterSize        = 16;
+	static constexpr size_t MeasurementBitWidth = 12;
+
+	/**
+	 * The minimum permissible Analogue-to-Digital Clock speed is 1 MHz, and the
+	 * maximum is 26 MHz, as per the data sheet.
+	 *
+	 * See table 65 on page 57:
+	 * https://docs.amd.com/v/u/en-US/ds181_Artix_7_Data_Sheet
+	 */
+	static constexpr size_t MinClockFrequencyHz = 1 * 1000 * 1000;
+	static constexpr size_t MaxClockFrequencyHz = 26 * 1000 * 1000;
+
+	/**
+	 * A clock divider value to be used by the Xilinx Analogue-to-Digital
+	 * Converter (XADC), which divides its input clock (the system clock)
+	 * by this divider to determine the clock of the XADC. This clock
+	 * must fall between specified maximum and minimum registerss, and the
+	 * divider must be an 8-bit integer, greater than 2.
+	 */
+	typedef uint8_t ClockDivider;
+
+	/**
+	 * Constructor - initialises the MMIO capability for the Analogue-to-
+	 * Digital converter, and then sets its clock divider and power down
+	 * mode.
+	 *
+	 * Takes the clock divider and power down modes to initialise. The
+	 * clock divider should divide the system clock to create a signal
+	 * between 1 and 26 MHz, and should be at least 2.
+	 */
+	SonataAnalogueDigitalConverter(ClockDivider  divider,
+	                               PowerDownMode powerDown)
+	{
+		registers = MMIO_CAPABILITY(uint32_t, adc);
```

Comment:
I think the overall approach is good, but this is an anti pattern. You are loading from the import table and then storing it in the memory here. Accessing the field is about as expensive as the original initialisation was. The typical pattern here is to add an always-inline function that wraps the macro.

Reply:
I see, thank you for the feedback - I see now the note in the CHERIoT Programmer's Guide about this anti-pattern. Am I understanding correctly that the inlining `MMIO_CAPABILITY` approach is more efficient because it expands to a single load capability load, whereas by loading once and storing to memory I am performing (more expensive) memory accesses equivalent to the original initialisation each time?

Reply:
Yes, the `MMIO_CAPABILITY` macro will expand to an auipcc (typically the compressed version since it will need a tiny offset) followed by a load-capability.  The other one may be a single load-capability relative to the base, but it now means that we have a stateful object and everything that entails.